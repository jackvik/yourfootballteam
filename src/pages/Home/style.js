import styled from "styled-components"; 

export const FlexStyle = styled.section`
    display: flex;
    flex-direction: column;
    width:45%;
`
export const TableWrapper = styled.div`
    max-height:80%;
    overflow-y:auto;
    table{
      width:100%;
    }
`
export const SectionTitle = styled.h2``;
export const FlexContainer = styled.div`
    height:calc(100vh - 7rem);
    display: flex;
    flex-direction: row;
    flex-basis: 45%;
    justify-content: space-between;
    margin:5px 15px;
    table {
      text-align: left;
      position: relative;
      border-collapse: collapse; 
}
button{
    height:fit-content;
}
th, td {
  padding: 0.25rem;
  text-align:center;
}
tr.red th {
  background: red;
  color: white;
}
tr.green th {
  background: green;
  color: white;
}
tr.purple th {
  background: purple;
  color: white;
}
th {
  background: white;
  position: sticky;
  top: 0; /* Don't forget this, required for the stickiness */
  box-shadow: 0 2px 2px -1px rgba(0, 0, 0, 0.4);
}
`