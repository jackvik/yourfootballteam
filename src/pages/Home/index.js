import React,{useState,useEffect} from 'react';
import {getData} from '../../api/ApiManager';
import {END_POINT_TEAMS_URL,END_POINT_PLAYERS_URL} from '../../api/APIEndPoints';
import {dispatchData, ValidationCheck} from '../../common/utils/utils';
import InputLabel from '@mui/material/InputLabel';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import Select from '@mui/material/Select';
import {SET_TEAMS_DATA,SET_MYTEAM_DATA,SET_PLAYERS_DATA} from '../../routes/action-types';
import PlayersView from '../../components/organisms/PlayersView';
import store from '../../routes/store';
import {FlexStyle,FlexContainer,SectionTitle,TableWrapper}  from './style';
import {useSelector} from 'react-redux';
import {ALL_PLAYERS_TITLE,MY_TEAM_SELECTION_TITLE,HomePageConstants} from '../../common/constants';
import Button from '@mui/material/Button';
function Home() {
  const [teamsData,setTeamsData] = useState([]);
  const [selectedCountry,setSelectedCountry] = useState("");
  const [teamPlayers,setTeamPlayers] = useState([]);
  let selectedPlayers = useSelector((state) => state.myTeamData.data);
  const [hasError, setAddError] = useState("");
  const [saveText, setSaveText] = useState("");
  useEffect(() => {
    getData(END_POINT_TEAMS_URL,successCallback,exceptionCallback)
  },[])

  const successCallback = (response) => {
    const dataPayload = {
      isSuccess: true,
      data: response.teams
    };

    setTeamsData(response.teams);
    dispatchData(SET_TEAMS_DATA,dataPayload);
  }
  
  const saveTeamHandler = () => {
    console.log("HEre");
    if (ValidationCheck(selectedPlayers,setAddError,setSaveText)) {
      console.log("Bingo")
      localStorage.setItem("my-team", JSON.stringify(selectedPlayers));
    }
  };
  const setTeamDropdownData = (data) => {
    return data?.sort((a, b) => (a.name > b.name ? 1 :b.name> a.name ? -1 : 0))
    .map((team) => {
      return (
        <MenuItem key={team.id} value={team.id}><img src={team.crestUrl} width={20} height={15} />{team.name}</MenuItem>
      )
    })
  }

  const exceptionCallback = (err) => {
    //console.log("error",error);
  }

  const handleChange = (event) => {
    setSelectedCountry(event.target.value);
    getData(END_POINT_PLAYERS_URL+event.target.value,getPlayersSuccessCallback,getPlayersErrorCallback);
  }

  const getPlayersSuccessCallback = (response) => {
    let playersResponse = response.squad.map((player) => ({
      ...player,country: response.name
    }));
    const dataPayload = {
      isSuccess: true,
      data: playersResponse,
    };
    setTeamPlayers(playersResponse);
    dispatchData(SET_PLAYERS_DATA,dataPayload);

  }
  const getPlayersErrorCallback = (err) => {

  }

  const addButtonHandler = (player) => {
    let selectedPlayersData = [...store.getState().myTeamData.data,player];
    let payload = {
      isSuccess: true,
      data: selectedPlayersData
    }

    dispatchData(SET_MYTEAM_DATA,payload);
  }

  const removeButtonHandler = (player) => {
    let modifiedData = [...selectedPlayers].filter((item) => item.id !== player.id);
    let payload = {
        isSuccess: true,
        data: modifiedData
    }
    dispatchData(SET_MYTEAM_DATA,payload);
}

  return (<>
      <FlexContainer>
        <FlexStyle>
        <SectionTitle>{ALL_PLAYERS_TITLE}</SectionTitle>
          <FormControl sx={{m:1,minWidth: 120}}>
            <InputLabel id={HomePageConstants.selectId}>{HomePageConstants.dropdownName}</InputLabel>
            <Select
            labelId={HomePageConstants.selectId}
            id={HomePageConstants.selectLabelel}
            value={selectedCountry}
            label={HomePageConstants.dropdownName}
            autoWidth
            onChange= {handleChange}
            >
              {setTeamDropdownData(teamsData)}
            </Select>
          </FormControl>
          {teamPlayers.length>0 ?<TableWrapper><PlayersView 
            isButton={true} 
            selectedPlayers={teamPlayers
              .filter(
                (player) => !selectedPlayers.map((sp) => sp.id).includes(player.id)
              )
              .sort((a, b) =>
                a.position > b.position ? 1 : b.position > a.position ? -1 : 0
              )} 
            clickHandler={addButtonHandler} 
            type={HomePageConstants.addBtn} 
            dataSelector={HomePageConstants.teamSelector}/>
            </TableWrapper> : <p>No data available</p>}
        </FlexStyle>
        <FlexStyle>
          <SectionTitle>{MY_TEAM_SELECTION_TITLE}</SectionTitle>
          <div>
            <Button 
              variant={HomePageConstants.btnVariant} 
              aria-label={HomePageConstants.saveBtn} 
              disabled={selectedPlayers.length===0} 
              onClick={saveTeamHandler}>
                {HomePageConstants.saveBtn}
            </Button>
          <p>{hasError}</p></div>
          <p>{saveText}</p>
          <TableWrapper>
            <PlayersView 
              isButton={true} 
              selectedPlayers={selectedPlayers} 
              clickHandler={removeButtonHandler} 
              type={HomePageConstants.rmvBtn} 
              dataSelector={HomePageConstants.myTeamSelector}/>
          </TableWrapper>  
        </FlexStyle>
      </FlexContainer>
    </>
  )
}

export default Home