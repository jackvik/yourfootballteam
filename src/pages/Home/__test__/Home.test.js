import React from 'react'
import { render} from "@testing-library/react"

import Home from '..'
import * as redux from 'react-redux'

const mockPlayer =[ {
    country: "Chelsea FC",
    id: "171",
    isButton : true,
    name: "Kai Havertz",
    position: "MidFielder",
  }];
  beforeEach(() => {
const spy = jest.spyOn(redux, 'useSelector')
spy.mockReturnValue(mockPlayer)

});

describe('<Home /> spec', () => {
    
it('renders the Home component', () => {
    const { asFragment } = render(<Home />)
    expect(asFragment(<Home />)).toMatchSnapshot()
 })



})