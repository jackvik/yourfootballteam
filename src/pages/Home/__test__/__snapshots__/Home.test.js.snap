// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> spec renders the Home component 1`] = `
<DocumentFragment>
  <div
    class="sc-eCImPb kSHbLi"
  >
    <section
      class="sc-gsDKAQ gZazSX"
    >
      <h5
        class="sc-hKwDye fQHpPY"
      >
        All Players
      </h5>
      <p>
        Loading...
      </p>
      <div
        class="sc-dkPtRN kOPygv"
      >
        <main
          class="sc-bdvvtL gbayWe"
        />
      </div>
    </section>
    <section
      class="sc-gsDKAQ gZazSX"
    >
      <h5
        class="sc-hKwDye fQHpPY"
      >
        My Team Players
      </h5>
      <div>
        <button
          aria-label="Save Team"
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          id="SaveBTN"
          tabindex="0"
          type="button"
        >
          Save Team
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <p />
      </div>
      <p />
      <div
        class="sc-dkPtRN kOPygv"
      >
        <main
          class="sc-bdvvtL gbayWe"
        >
          <table>
            <thead>
              <tr>
                <th>
                  Team
                </th>
                <th>
                  Position
                </th>
                <th>
                  Name
                </th>
                <th>
                  Action
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  Chelsea FC
                </td>
                <td>
                  MID
                </td>
                <td>
                  Kai Havertz
                </td>
                <td>
                  <button
                    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    REMOVE
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </main>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
